pyteal/ast/router.py

```diff
@@ -459,34 +566,90 @@ def wrap_handler(
                     abi.MethodReturn(output_temp),
                     Approve(),
                 )
+        else:
+            subroutine_ify = ASTBuilder.__de_abify_subroutine(handler)
+            return Seq(
+                subroutine_ify(),
+                Approve(),
+            )
+
+    @staticmethod
+    def __de_abify_subroutine(
+        handler: ABIReturnSubroutine | SubroutineFnWrapper | Expr,
+    ) -> SubroutineFnWrapper:
+        handler = ASTBuilder.__handler_validity_check(handler)
+        (
+            arg_vals,
+            app_arg_vals,
+            txn_arg_vals,
+        ) = ASTBuilder.__subroutine_argument_instance_generate(handler)
+
+        (
+            decode_instructions,
+            arg_vals,
+            proto,
+        ) = ASTBuilder.__decode_constructions_and_args(
+            arg_vals,
+            app_arg_vals,
+            txn_arg_vals,
+            handler,
+            use_frame_pt=True,
+        )
+
+        subroutine_caster = Subroutine(TealType.none, f"{handler.name()}_caster")
+        return_expr: Expr
+
+        if not proto or not proto.mem_layout:
+            raise TealInternalError(
+                "Assuming to be using frame pointer, memory layout and proto should be available."
+            )
+
+        if handler.type_of() == sdk_abi.Returns.VOID:
+            return_expr = Seq(
+                *decode_instructions,
+                cast(Expr, handler(*arg_vals)),
+            )
+        else:
+            output_temp: abi.BaseType = cast(
+                OutputKwArgInfo, handler.output_kwarg_info
+            ).abi_type.new_instance()
+            output_temp._stored_value = FrameVar(proto, 0)
+            subroutine_call: abi.ReturnedValue = cast(
+                abi.ReturnedValue, handler(*arg_vals)
+            )
+
+            return_expr = Seq(
+                *proto.mem_layout._succinct_repr(),
+                *decode_instructions,
+                subroutine_call.store_into(output_temp),
+                abi.MethodReturn(output_temp),
+            )
+
+        def declaration():
+            return return_expr
+
+        return subroutine_caster(declaration)
 
     def add_method_to_ast(
         self, method_signature: str, cond: Expr | int, handler: ABIReturnSubroutine
     ) -> None:
-        walk_in_cond = Txn.application_args[0] == MethodSignature(method_signature)
-        match cond:
-            case Expr():
-                self.conditions_n_branches.append(
-                    CondNode(
-                        walk_in_cond,
-                        Seq(Assert(cond), self.wrap_handler(True, handler)),
-                    )
-                )
-            case 1:
-                self.conditions_n_branches.append(
-                    CondNode(walk_in_cond, self.wrap_handler(True, handler))
-                )
-            case 0:
```

Comment:
noticed this one has a condition for 0, the new one doesnt?

Reply:
The case has been ruled out in the new code snippet
```python
        if isinstance(cond, int) and cond == 0:
            return
```